### PR TITLE
Custom Enumerables

### PR DESCRIPTION
**Pull request by KennyUTN**
The following is a recreation of the Enumerable library on Ruby!

the following methods were remade. 

> Each
> Each with index
> Select
> All?
> Any?
> None?
> Count
> Map
> Inject

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,61 @@
...
+
+- Open related file by using text editor.
+
+## Author
+👤 **Juan Ignacio Asis**
+
+- Github: [@KennyUTN](https://github.com/KennyUTN)
+- Twitter: [@Kenny11614766](https://twitter.com/Kenny11614766)
+- Linkedin: [linkedin](https://www.linkedin.com/in/ignacio-asis-b8214b183/)
+
```
Comment (javitocor) [inline README.md]: In the LinkedIn field, kindly add your name, the link should not be 'linkedin'

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  end
+
+  def my_all?
+    return nil unless block_given?
+
+    my_each do |elem|
+      return false unless yield elem
+    end
+    true
+  end
```
Comment (javitocor) [inline main.rb]: - enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
- enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- enumerables #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
- enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- enumerables #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_any?
+    return nil unless block_given?
+
+    my_each do |elem|
+      return true if yield elem
+    end
+    false
+  end
```
Comment (javitocor) [inline main.rb]: - enumerables #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true
- enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
- enumerables #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- enumerables #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_none?
+    return nil unless block_given?
+
+    my_each do |elem|
+      return false if yield elem
+    end
+    true
+  end
```
Comment (javitocor) [inline main.rb]: - enumerables #my_none? when no block or argument is given returns true only if none of the collection members is true
- enumerables #my_none? when no block or argument is given returns false only if one of the collection members is true
- enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- enumerables #my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
- enumerables #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
- enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
- enumerables #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_count
+    return size unless block_given?
+
+    n = 0
+    my_each do
+      n += 1
+      n
+    end
+  end
```
Comment (javitocor) [inline main.rb]: - enumerables #my_count counts the number of items in enum that are equal to item if an argument is given   
- enumerables #my_count counts the number of elements yielding a true value if a block is given

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    return dup if proc.nil?
+
+    ary = []
+    if proc.nil?
+      to_a.my_each { |elem| ary << yield(elem) }
+    else
+      to_a.my_each { |elem| ary << proc.call(elem) }
+    end
+    ary
+  end
```
Comment (javitocor) [inline main.rb]: - enumerables #my_map returns a new array with the results of running a given block once for every element in enum.
- enumerables #my_map returns a new array with the results of running a given block once for every element in enum.
- enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range
- enumerables #my_map returns an Enumerator if no block is given
- enumerables #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      acc
+    elsif !sim.nil? && sim.is_a?(Symbol)
+      acc = arg
+      my_each { |i| acc = acc.nil? ? i : acc.send(sim, i) }
+      acc
+    else
+      yield
+    end
+  end
+end
```
Comment (javitocor) [inline main.rb]: - enumerables #my_inject raises a "LocalJumpError" when no block or argument is given
- enumerables #my_inject searches for the longest word in an array of strings
- enumerables #my_inject does not mutate the original array
- enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+def multiply_els(array)
+  array.my_inject(:*)
+end
```
Comment (javitocor) [inline main.rb]: - enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

Hello! ✋ 
Good job so far in this project! 👍 I know it is challenging to create all those methods, but you did it quite well! 💯  
However, we still need to implement some changes in order to get the project approved 💪 
Kindly check my comments below 👇 apply the changes and resubmit your project for a new code review ⏩ 
Good luck! 😃 
## Mandatory changes
- Kindly make sure your code pass all the linters check, now this is not happening, kindly fix it
- The live link to repl.it in the readme file is not working, kindly fix it
- In your readme file, improve your set up section making clear all the steps to follow in order, think a visitor with no previous experience should be able to run your project. So try to be as much specific as you can.
- You can find more info about the enumerables, here https://ruby-doc.org/core-2.6.5/Enumerable.html

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @KennyUTN  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] The license link, theodinproject link, and issues link are not working in your readme file, fix them please
- [ ] Make sure that you fix all of the issues mentioned  in the previous review, almost all test cases the previous TSE mentioned, are still failing. Please fix them before you request for re-review
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by Georjane — APPROVED** [review]
## Status: PR Approved 💯 🥇 
### Hi Team ✋🏽 
Congratulations on successfully completing the Enumerable Project. Your methods pass all test cases. It's a thumbs up for me. 👍🏽  You can go ahead and merge this pull request.

Good luck on your next project and Happy Coding 💻

**Pull request merged**